kernel/src/net/socket/ip/stream/mod.rs

```diff
@@ -719,17 +751,61 @@ impl SocketEventObserver for StreamSocket {
     }
 }
 
-impl Drop for StreamSocket {
-    fn drop(&mut self) {
-        let state = self.state.write().take();
+impl State {
+    /// Calls `f` to set raw socket option.
+    ///
+    /// Note that for listening socket, `f` is called on all backlog sockets in `Listen` State.
+    /// That is to say, `f` won't be called on backlog sockets in `SynReceived` or `Established` state.
+    fn set_raw_option(&mut self, set_option: impl Fn(&mut dyn RawTcpSetOption)) {
+        match self {
+            State::Init(init_stream) => init_stream.set_raw_option(set_option),
+            State::Connecting(connecting_stream) => connecting_stream.set_raw_option(set_option),
+            State::Connected(connected_stream) => connected_stream.set_raw_option(set_option),
+            State::Listen(listen_stream) => listen_stream.set_raw_option(set_option),
+        }
+    }
 
-        let iface_to_poll = match state {
+    fn iface_to_poll(&self) -> Option<Arc<Iface>> {
+        match self {
             State::Init(_) => None,
             State::Connecting(ref connecting_stream) => Some(connecting_stream.iface().clone()),
             State::Connected(ref connected_stream) => Some(connected_stream.iface().clone()),
             State::Listen(ref listen_stream) => Some(listen_stream.iface().clone()),
+        }
+    }
+}
+
+impl SetSocketLevelOption for State {
+    fn set_keep_alive(&mut self, keep_alive: bool) -> bool {
```

Comment:
Instead of changing this method's signature to return boolean, I would rather do this:
```rust
pub trait RawTcpSetOption {
    /// Sets the keepalive interval.
    ///
    /// Polling the iface _may_ be required after this method succeeds.
    fn set_keep_alive(&mut self, interval: Option<Duration>) -> NeedIfacePoll;
}
```
This is consistent with `BoundTcpSocket::{send,recv}`, which also returns `NeedIfacePoll` to indicate a conditionally required poll.
https://github.com/asterinas/asterinas/blob/fc45592a28b0cdfa0816fd58a7d2a51105532b68/kernel/libs/aster-bigtcp/src/socket/bound.rs#L318-L321

Then, `SetSocketLevelOption::set_keep_alive` will also return `NeedIfacePoll`. Following that, `SocketOptionSet::set_option` should return `Result<NeedIfacePoll>`. Note that comments describing what the returned boolean means can be safely removed (as the type name is clear enough).